docs/pages/tutorials/003_cuda_spmv.md

```diff
@@ -0,0 +1,274 @@
+// to the correct type in the same call to `Permute1D`
+Array<value_type>* gray_reordered_vec = ReorderBase::Permute1D<Array>(gray_reorder, vec, {&cpu}, true);
+// We move this array to the GPU.
+cuda::CUDAArray<value_type>* cuda_gray_reordered_vec = gray_reordered_vec->Convert<cuda::CUDAArray>(&gpu0);
+```
+
+The `Permute2D` call returns a `CSR` object, but it returns it as a polymorphic pointer to the superclass `FormatOrderTwo` which is the parent of `CSR` and other order-2 formats. We can directly copy the returned `CSR` to the GPU through this polymorphic pointer.
+
+Similar to `Permute2D`, `Permute1D` also returns a polymorphic pointer at the superclass `FormatOrderOne` (parent of `Array` and `CUDAArray`). However, if we know the data type of the returned object, we can cast the returned pointer to that type by passing it as a template argument to the function. Note that, if the cast fails, an exception will be thrown.
+
+Let's try multiplication using the reordered data.
+```c++
+start = std::chrono::system_clock::now();
+spmv_1w1r<<<NUM_BLOCKS, THREADS_PER_BLOCK>>>(cuda_gray_reordered_csr ->get_row_ptr(),
+                                       cuda_gray_reordered_csr->get_col(),
+                                       cuda_gray_reordered_csr->get_vals(),
+                                       cuda_gray_reordered_vec->get_vals(),
+                                       result_ptr, num_rows);
+cudaDeviceSynchronize();
+if (!cudaPeekAtLastError() == cudaSuccess){
+  std::cout << "Kernel failed: " << cudaGetErrorString(cudaPeekAtLastError()) << std::endl;
+  return 1;
+}
+end = std::chrono::system_clock::now();
+total_time = end-start;
+std::cout << "SpMV with Gray reordering takes: " << total_time.count() << " seconds" << std::endl;
+```
+
+### 6. Generate RCM reordering of a matrix by utilizing automatic function matching and conversion. 
+Now, let's try Reverse Cuthill McKee (RCM) reordering but this time, take advantage of the function matching capability in SparseBase.
+
+```c++
+// We can pass the CUDACSR and the function will automatically convert it to CSR for reordering
+id_type* rcm_reorder = ReorderBase::Reorder<RCMReorder>({}, cuda_csr, {&cpu}, true);
+```
+Every preprocessing algorithm in the library (reordering, partitioning, feature extraction, etc.) can be implemented for many format types. In this case, `RCMReorder`is only implemented for the `CSR` type. However, when we set the `convert_input` parameter (last parameter) to `true` in the call, this will allow function matching to take place. 
+
+Function matching takes the input formats to a preprocessing and the contexts the user passes to the function call, and, if the input format can't be used directly for the preprocessing (i.e., no function exists for the input's format type), it attempts to convert the input (using the passed contexts) to a format for which an implementation exists in the preprocessing. 
```

Comment:
Maybe specifically mention what conversion via the context means. So in this case the CSR actually gets copied to the CPU.

Reply:
Makes sense.